zerver/views/streams.py

```diff
@@ -75,12 +77,69 @@
 from zerver.lib.typed_endpoint import ApiParamConfig, PathOnly, typed_endpoint
 from zerver.lib.typed_endpoint_validators import check_color, check_int_in_validator
 from zerver.lib.user_groups import access_user_group_for_setting
-from zerver.lib.users import access_user_by_email, access_user_by_id
+from zerver.lib.users import access_user_by_email, access_user_by_id, access_user_common
 from zerver.lib.utils import assert_is_not_none
 from zerver.models import NamedUserGroup, Realm, Stream, UserProfile
 from zerver.models.users import get_system_bot
 
 
+def bulk_principals_to_user_profiles(
+    agent: UserProfile, principals: list[str] | list[int]
+) -> set[UserProfile]:
+    # principals is guaranteed to be non-empty and to have the same type of elements, int or str,
+    # so the following check is safe and enough.
+
+    # principlas are user emails.
+    if isinstance(principals[0], str):
+        # using .lower() for each email and
+        # annotate(lower_email=Lower("email")) in the queryset
+        # achieves the same result of doing email__iexact__in=emails but that is not supported.
+        user_emails = [email.strip().lower() for email in principals]  # type: ignore[union-attr] # email is str as explained above
```

Comment:
Call this variable `lower_user_emails`; that'll make reading the rest a lot clearer.

Reply:
ok

Reply:
> Hmm but it might be that you want `upper`, not lower; that's what we have database indexes on, from `\d zerver_userprofile` in `manage.py dbshell`:

So we won't need to do `annotate(email_upper=Upper("email"))` ?
Because the field itself is not upper but its index is upper so the filter will check against the index only.

Am I understanding correctly ?